versions/development/SPEC.md

```diff
@@ -1544,6 +1545,78 @@ workflow foo {
 }
 ```
 
+### Environment Variables
+
+Any input to a task may be converted into an environment variable that will be accessible from within the task's 
+shell environment during command execution. Unlike inputs, environment variables are not templated into the command when
+preparing the execution script, but are actually available at runtime directly from the environment. They may be accessed using
+normal shell variable access semantics (ie `$FOO` or `${FOO}`).
+
+In order to access an input value as an environment variable, you can add the `env` modifier preceding a declaration anywhere
+within the task. This applies to declarations within the input section as well as [private declarations](#private-declarations).
+An important thing to note is that while the `env` modifier makes the declaration available in the
+environment, it does not change its access to normal WDL expressions within a command. That means you can refer to a declaration
+annotated with `env` either through the shell semantics (`${FOO}`) or through normal WDL semantics (`~{FOO}`).
+
+When an input is annotated with `env` it is the engines responsibility to serialize the value appropriately into a string.
+Serializing complex values (`Map`, `Pair`, `Struct`, `Array`) should be equivalent to calling `write_json` on the given
```

Comment:
Add a link `[WDL Value Serialization and Deserialization](#appendix-a-wdl-value-serialization-and-deserialization)`

Reply:
Great point to discuss:

On the one hand, if the developer definitely wants the JSON in a file then they can easily add a private declaration `env File x_json = write_json(x)`. Given that's easy, maybe the environment variable for a compound value should contain the JSON directly.

On the other hand, JSON directly in the environment variable isn't very natural to work with in shell; often you'd end up just dumping it into a file to be parsed anyway. Also, developers should be discouraged from from cramming large compound values directly into environment variables, because there will usually be some system-imposed limit on environment size.

TBH, `env File x_json = write_json(x)` makes the intent much clearer to somebody reading the code, so whichever way we go on this question, I think we should actually discourage using `env` with compound values at all.

Reply:
@mlin JSON directly as environment variables is actually not as cumbersome as you may think. We regularly use it in our own CI/CD infrastructure  and pass values to and from tools like `jq` which are great at parsing json. We rarely ever need to use temporary files and just pipe input from one command to the other, storing intermediate entries in a environment variable.

I think pointing to the example of the `write_json` is actually not an accurate representation of what is happening. we probably want to just say that complex types will be serialized to `JSON` prior to being set as an environment variable.

There is of course the limitation on size of the environment variables. So Maybe we can provide guidance here: "Engines may impose limits on the total length a single environment variable is allowed to occupy as well as the number of environment variables that are allowed to be passed into a single task. If such limitations exist, engines should clearly document them for the end user."